libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1729,31 +1730,80 @@ bool TypeChecker::visit(UnaryOperation const& _operation)
 		requireLValue(_operation.subExpression(), false);
 	else
 		_operation.subExpression().accept(*this);
-	Type const* subExprType = type(_operation.subExpression());
-	TypeResult result = type(_operation.subExpression())->unaryOperatorResult(op);
-	if (!result)
+	Type const* operandType = type(_operation.subExpression());
+
+	// Check if the operator is built-in or user-defined.
+	TypeResult builtinResult = operandType->unaryOperatorResult(op);
+	set<FunctionDefinition const*> matchingDefinitions = operandType->operatorDefinitions(
+		op,
+		*currentDefinitionScope(),
+		true // _unary
+	);
+
+	// Operator can't be both user-defined and built-in at the same time.
+	solAssert(!builtinResult || matchingDefinitions.empty());
+
+	// By default use the type we'd expect from correct code. This way we can continue analysis
+	// of other expressions in a sensible way in case of a non-fatal error.
+	Type const* resultType = operandType;
+
+	FunctionDefinition const* operatorDefinition = nullptr;
+	if (builtinResult)
+		resultType = builtinResult;
+	else if (!matchingDefinitions.empty())
+	{
+		if (matchingDefinitions.size() >= 2)
+		{
+			SecondarySourceLocation secondaryLocation;
+			for (FunctionDefinition const* definition: matchingDefinitions)
+				secondaryLocation.append("Candidate definition:", definition->location());
```

Comment:
Oh wow, this has nondeterministic order, doesn't it?

Reply:
True. Good catch. Should be fixed now (did not build locally for me yet but CI will catch it if it's not).

I tried to go for AST ID ordering at first but then I'd have to change some includes from `ASTForward.h` to `AST.h` because `ASTNode::CompareByID` is nested and can't be forward declared. I guess a `vector` is fine here. Fortunately we don't really need to look up stuff in it. We only check how many definitions it has and sometimes enumerate them.